compositor_render/src/renderer/common_pipeline/surface.rs

```diff
@@ -0,0 +1,226 @@
+        while i < N / 6 {
+            indices[6 * i] = (4 * i) as u16;
+            indices[6 * i + 1] = (4 * i + 1) as u16;
+            indices[6 * i + 2] = (4 * i + 2) as u16;
+            indices[6 * i + 3] = (4 * i + 2) as u16;
+            indices[6 * i + 4] = (4 * i + 3) as u16;
+            indices[6 * i + 5] = (4 * i) as u16;
+            i += 1;
+        }
+
+        Self(indices)
+    }
+}
+
+/// Abstraction for buffers, holding vertices and indices of
+/// 2D planes, on which textures are rendered
+pub struct Surfaces {
+    inputs_vertices: Buffer,
+    inputs_indices: Buffer,
+    no_inputs_vertices: Buffer,
+    no_inputs_indices: Buffer,
+}
+
+const VERTICES_COUNT: usize = 4 * MAX_TEXTURES_COUNT as usize;
+const INDICES_COUNT: usize = 6 * MAX_TEXTURES_COUNT as usize;
+
+/// Vertex and index buffer that describe render area as an rectangle mapped to texture.
+impl Surfaces {
+    /// Vertices of texture 2D planes passed to the vertex shader.
+    /// Each plane has 4 vertices
+    const ALL_PLANES_VERTICES: Vertices<VERTICES_COUNT> = Vertices::new();
+
+    /// Indexes vertices of texture 2D planes passed to vertex shader.
+    /// Describes which vertices combine triangles.
+    /// Each texture plane contain 2 triangles - 6 indices
+    const ALL_PLANES_INDICES: Indices<INDICES_COUNT> = Indices::new();
+
+    pub fn new(device: &wgpu::Device) -> Self {
+        let inputs_vertex_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+            label: Some("vertex buffer"),
+            usage: wgpu::BufferUsages::VERTEX | wgpu::BufferUsages::COPY_DST,
+            contents: bytemuck::cast_slice(&Self::ALL_PLANES_VERTICES.0),
+        });
+
+        let inputs_index_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+            label: Some("index buffer"),
+            usage: wgpu::BufferUsages::INDEX | wgpu::BufferUsages::COPY_DST,
+            contents: bytemuck::cast_slice(&Self::ALL_PLANES_INDICES.0),
+        });
+
+        let no_inputs_vertex_buffer =
+            device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+                label: Some("vertex buffer"),
+                usage: wgpu::BufferUsages::VERTEX | wgpu::BufferUsages::COPY_DST,
+                contents: bytemuck::cast_slice(&SINGLE_PLANE_VERTICES),
+            });
+
+        let no_inputs_index_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+            label: Some("index buffer"),
+            usage: wgpu::BufferUsages::INDEX | wgpu::BufferUsages::COPY_DST,
+            contents: bytemuck::cast_slice(&SINGLE_PLANE_INDICES),
+        });
+
+        Self {
+            inputs_vertices: inputs_vertex_buffer,
+            inputs_indices: inputs_index_buffer,
+            no_inputs_vertices: no_inputs_vertex_buffer,
+            no_inputs_indices: no_inputs_index_buffer,
+        }
+    }
+
+    pub fn draw<'a>(&'a self, render_pass: &mut wgpu::RenderPass<'a>, input_textures_count: u32) {
```

Comment:
Is this the correct name? if I understand correctly nothing here is triggering drawing itself, Is it?

Reply:
it doesn't draw. we could call this  `encode_draw`, but at least to me just `draw` is fine